### PR TITLE
Moonriver erc20 NativeCoin contracts

### PR DESCRIPTION
**Pull request by simsonraj**

### CODE REVIEW

File: docker-compose/goloop2moonriver/README.MD [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+cd ~/work/btp
+git fetch
+git checkout icondao
+make btpsimple-image
+```
+
+### Run docker-compose
+
+```
+cd ~/work/btp/docker-compose/goloop2moonbeam
```
Comment (danielvo11) [inline docker-compose/goloop2moonriver/README.MD]: goloop2moonriver???

File: solidity/nativecoinERC20/contracts/BSHCore.sol [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+
+/**
+   @title BSHCore contract
+   @dev This contract is used to handle coin transferring service
+   Note: The coin of following contract can be:
+   Native Coin : The native coin of this chain
+   Wrapped Native Coin : A tokenized ERC1155 version of another native coin like ICX
```
Comment (danielvo11) [inline solidity/nativecoinERC20/contracts/BSHCore.sol]: please update description

File: solidity/nativecoinERC20/contracts/BSHCore.sol [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+        //  Before that, Caller must approve (setApproveForAll) to accept
+        //  token being transfer out by an Operator
+        //  If this requirement is failed, a transaction is reverted.
+        //  After transferring token, BSHCore contract updates Caller's locked balance
+        //  as a record of pending transfer transaction
+        //  When a transaction is completed without any error on another chain,
+        //  Locked Token amount (bind to an address of caller) will be reset/subtract,
+        //  then emit a successful TransferEnd event as a notification
+        //  Otherwise, the locked amount will also be updated
+        //  but BSHCore contract will issue a refund to Caller before emitting an error TransferEnd event
```
Comment (danielvo11) [inline solidity/nativecoinERC20/contracts/BSHCore.sol:334]: Please update description

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+    }
+
+    @External
+    public void reclaim(String _coinName, BigInteger _value) {
+        require(_value.compareTo(BigInteger.ZERO) > 0, "_value must be positive");
+
+        Address owner = Context.getCaller();
+        Balance balance = getBalance(_coinName, owner);
+        require(balance.getRefundable().compareTo(_value) > -1, "invalid value");
+        balance.setRefundable(balance.getRefundable().subtract(_value));
```
Comment (danielvo11) [inline javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: They can not claim usable balance?

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+        } else {
+            transfer(irc2, owner, _value, "Reclaim: transferring to Caller".getBytes());
+        }
+    }
+
+    @External
+    public void tokenFallback(Address from, BigInteger value, @Optional byte[] data) {
+        require(value.compareTo(BigInteger.ZERO) > 0, "Invalid Amount");
+        require(Context.getCaller().equals(irc2), "Invalid Caller: Token Contract wrong");
+        require(tokenName != null && !tokenName.equals(""), "Token not registered");
```
Comment (danielvo11) [inline javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: what is the purpose of this line?

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+        }
+    }
+
+    private Balance getBalance(String coinName, Address owner) {
+        Balance balance = balances.at(coinName).get(owner);
+        if (balance == null) {
+            balance = new Balance(BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO);
+            return balance;
+        }
+        Balance newBal = new Balance(balance.getUsable(), balance.getLocked(), balance.getRefundable());
```
Comment (danielvo11) [inline javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java:332]: Is it necessary to new Balance instance here?

**Comment by danielvo11** [comment]
I check that most of code in `goloop2moonriver` are copied from `goloop2moobeam`, why do you need to create new folder? I think you just need to add `nativeCoinIRC2BSH` service to current `goloop2moobeam` folder.

File: docker-compose/goloop2moonriver/README.MD (hunk 1 above)
Comment (simsonraj) [reply docker-compose/goloop2moonriver/README.MD]: Now the moonriver folder is removed & merged with moonbeam as suggested

File: solidity/nativecoinERC20/contracts/BSHCore.sol (hunk 2 above)
Comment (simsonraj) [reply solidity/nativecoinERC20/contracts/BSHCore.sol]: Thanks for pointing it out, updated this.

File: solidity/nativecoinERC20/contracts/BSHCore.sol (hunk 3 above)
Comment (simsonraj) [reply solidity/nativecoinERC20/contracts/BSHCore.sol:334]: Thanks for pointing it out, updated this.

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java (hunk 4 above)
Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: Thanks for pointing it out, fixed this issue.

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java (hunk 5 above)
Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: ERC20 in javascore handles token deposit to contract using tokenFallback mechanism

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java (hunk 6 above)
Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java:332]: While doing javascore unit testing, without the creation of new instance, it updated all previous instances of balances in the test lines while executing. I wasn't able to exactly pointout why, but this fixed the issue

**Comment by simsonraj** [comment]
> I check that most of code in `goloop2moonriver` are copied from `goloop2moobeam`, why do you need to create new folder? I think you just need to add `nativeCoinIRC2BSH` service to current `goloop2moobeam` folder.

I started with a new folder initially anticipating more changes that might have conflicts, now that it's clear, I merged it back to moonbeam folder.

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java (hunk 5 above)
Comment (danielvo11) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: are there any cases that tokenName can be null?

Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: Good catch, this line was carried over from multi token support from Token-transfer BSH. removed it now. Thanks

File: javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java [hunk 7]
```diff
@@ -0,0 +1,583 @@
...
+
+    @External
+    public void reclaim(String _coinName, BigInteger _value) {
+        require(_value.compareTo(BigInteger.ZERO) > 0, "_value must be positive");
+
+        Address owner = Context.getCaller();
+        Balance balance = getBalance(_coinName, owner);
+
+        if (balance.getUsable().compareTo(BigInteger.ZERO) > 0) {
+            balance.setRefundable(balance.getRefundable().add(balance.getUsable()));
```
Comment (danielvo11) [inline javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: So after claim, all my usable balance added to refundable balance, and then if I want to transfer, I need to deposit again. Is it better to convert refundable balance to usable balance.

Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: so there will be a new method to claim usable balance?

Comment (danielvo11) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: I mean add refundable balance to usable balance and then subtract usable balance by value user want to claim. I make  usable balance still remain, user don't need to deposit it again.

Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: i see, i think that's a better idea too. ill do it

Comment (simsonraj) [reply javascore/nativecoinIRC2/src/main/java/foundation/icon/btp/nativecoinIRC2/NativeCoinService.java]: This is Done!

**Comment by danielvo11** [comment]
LGTM, @quocle108 please merge it

**Pull request merged**